chain/vm/execution.go

```diff
@@ -0,0 +1,227 @@
+	DefaultAvailableExecutionLanes = 4
+	// DefaultPriorityExecutionLanes is the number of reserved execution lanes for priority computations.
+	// This is purely userspace, but we believe it is a reasonable default, even with more available
+	// lanes.
+	DefaultPriorityExecutionLanes = 2
+)
+
+var ErrExecutorDone = errors.New("executor has been released")
+
+// the execution environment; see below for definition, methods, and initilization
+var execution *executionEnv
+
+// implementation of vm executor with simple sanity check preventing use after free.
+type vmExecutor struct {
+	lk    sync.RWMutex
+	vmi   Interface
+	token *executionToken
+	done  bool
+}
+
+var _ Executor = (*vmExecutor)(nil)
+
+func newVMExecutor(vmi Interface, token *executionToken) Executor {
+	return &vmExecutor{vmi: vmi, token: token}
+}
+
+func (e *vmExecutor) ApplyMessage(ctx context.Context, cmsg types.ChainMsg) (*ApplyRet, error) {
+	e.lk.RLock()
+	defer e.lk.RUnlock()
+
+	if e.done {
+		return nil, ErrExecutorDone
+	}
+
+	return e.vmi.ApplyMessage(ctx, cmsg)
+}
+
+func (e *vmExecutor) ApplyImplicitMessage(ctx context.Context, msg *types.Message) (*ApplyRet, error) {
+	e.lk.RLock()
+	defer e.lk.RUnlock()
+
+	if e.done {
+		return nil, ErrExecutorDone
+	}
+
+	return e.vmi.ApplyImplicitMessage(ctx, msg)
+}
+
+func (e *vmExecutor) Flush(ctx context.Context) (cid.Cid, error) {
+	e.lk.RLock()
+	defer e.lk.RUnlock()
+
+	if e.done {
+		return cid.Undef, ErrExecutorDone
+	}
+
+	return e.vmi.Flush(ctx)
+}
+
+func (e *vmExecutor) Done() {
+	e.lk.Lock()
+	defer e.lk.Unlock()
+
+	if !e.done {
+		e.token.Done()
+		e.token = nil
+		e.done = true
+	}
+}
+
+type executionToken struct {
+	lane     ExecutionLane
+	reserved int
+}
+
+func (token *executionToken) Done() {
+	execution.putToken(token)
+}
+
+type executionEnv struct {
+	mx   *sync.Mutex
+	cond *sync.Cond
+
+	// available executors
+	available int
+	// reserved executors
+	reserved int
+}
+
+func (e *executionEnv) getToken(lane ExecutionLane) *executionToken {
+	metricsUp(metrics.VMExecutionWaiting, lane)
+	defer metricsDown(metrics.VMExecutionWaiting, lane)
+
+	e.mx.Lock()
+	defer e.mx.Unlock()
+
+	switch lane {
+	case ExecutionLaneDefault:
+		for e.available <= e.reserved {
+			e.cond.Wait()
+		}
+
+		e.available--
+
+		metricsUp(metrics.VMExecutionRunning, lane)
+		return &executionToken{lane: lane, reserved: 0}
+
+	case ExecutionLanePriority:
+		for e.available == 0 {
+			e.cond.Wait()
+		}
+
+		e.available--
+
+		reserving := 0
+		if e.reserved > 0 {
+			e.reserved--
+			reserving = 1
+		}
+
+		metricsUp(metrics.VMExecutionRunning, lane)
+		return &executionToken{lane: lane, reserved: reserving}
+
+	default:
+		// already checked at interface boundary in NewVM, so this is appropriate
+		panic("bogus execution lane")
+	}
+}
+
+func (e *executionEnv) putToken(token *executionToken) {
+	e.mx.Lock()
+	defer e.mx.Unlock()
+
+	e.available++
+	e.reserved += token.reserved
+
+	e.cond.Broadcast()
```

Comment:
I think a comment would be useful here explaining why we need to use `Broadcast` and not `Signal`. This isn't obvious to me, after all we are putting back one token 🤔

Reply:
this is pretty standard; plus it is two lanes, so signal would be incorrect.

But really, you almost always use Broadcast; Signal is the unusual one as you need to prove there is exactly one thread that should make progress.
I dont think this needs a comment, but if i get more people wanting a comment I will add one.

Reply:
> there is exactly one thread that should make progress

Most probably I misunderstood `Signal` then. If this code is pretty standard in Go then it's all fine.

Reply:
added comment.